Firestore/core/src/firebase/firestore/model/document_map.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_DOCUMENT_MAP_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_DOCUMENT_MAP_H_
+
+#include <utility>
+
+#import "Firestore/Source/Model/FSTDocument.h"
+
+#include "Firestore/core/src/firebase/firestore/immutable/sorted_map.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+
+#include "absl/base/attributes.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+/**
+ * Convenience type for a map of keys to MaybeDocuments, since they are so
+ * common.
+ */
+using MaybeDocumentMap = immutable::SortedMap<DocumentKey, FSTMaybeDocument*>;
+
+/**
+ * Convenience type for a map of keys to Documents, since they are so common.
+ *
+ * PORTING NOTE: unlike other platforms, in C++ `Foo<Derived*>` cannot be
+ * converted to `Foo<Base*>`; consequently, if `DocumentMap` were simply an
+ * alias similar to `MaybeDocumentMap`, it couldn't be passed to functions
+ * expecting `MaybeDocumentMap`.
+ *
+ * To work around this, in C++ `DocumentMap` is a simple wrapper over
+ * a `MaybeDocumentMap` that forwards all functions to the underlying map but
+ * with added type safety (it only accepts `FSTDocument`s, not
+ * `FSTMaybeDocument`s). Use `DocumentMap` in functions creating and/or
+ * returning maps that only contain `FSTDocument`s; when the `DocumentMap` needs
+ * to be passed to a function accepting a `MaybeDocumentMap`, use
+ * `underlying_map` function to get (read-only) access to the representation.
+ * Also use `underlying_map` for iterating and searching.
+ */
+class DocumentMap {
+ public:
+  DocumentMap() = default;
+
+  ABSL_MUST_USE_RESULT DocumentMap insert(const DocumentKey& key,
+                                          FSTDocument* value) const {
+    return DocumentMap{map_.insert(key, value)};
+  }
+
+  ABSL_MUST_USE_RESULT DocumentMap erase(const DocumentKey& key) const {
+    return DocumentMap{map_.erase(key)};
+  }
+
+  bool empty() const {
+    return map_.empty();
+  }
+  MaybeDocumentMap::size_type size() const {
+    return map_.size();
+  }
+
+  /** Use this function to "convert" `DocumentMap` to a `MaybeDocumentMap`. */
+  const MaybeDocumentMap& underlying_map() const {
```

Comment:
Elsewhere we've used asFoo for the name of a method that returned a Foo-typed view of some other type's underlying data. In that vein, maybe `as_maybe_documents`? `as_maybe_document_map`?

It's a little weird in this case though because the accessor-naming pattern makes the actual name seem unrelated to the type.

I have no strong feelings on this.

Reply:
Hmm, I cannot find any names by regex ` as[A-Z]`, so I can't compare, but to me, `as` sort of implies creating new object (I take it to mean something like "create a version of this object represented as a different type). I think `underlying` implies that a reference is returned, so that it's clear it's a cheap operation. I also don't feel at all strongly about this; I'll be happy to rename if you'd like.